include/circt/Scheduling/Problems.h

```diff
@@ -231,6 +231,40 @@ class Problem {
   LogicalResult verify();
 };
 
+/// This class models a cyclic scheduling problem. Its solution solution can be
+/// used to construct a pipelined datapath with a fixed, integer initiation
+/// interval, in which the execution of multiple iterations/samples/etc. may
+/// overlap.
+class CyclicProblem : public virtual Problem {
+private:
+  DependenceProperty<unsigned> distance;
+  ProblemProperty<unsigned> initiationInterval;
+
+public:
+  using Problem::Problem;
+
+  /// The distance determines whether a dependence has to be satisfied in the
+  /// same iteration (distance=0 or not set), or distance-many iterations later.
+  Optional<unsigned> getDistance(Dependence dep) {
+    return distance.lookup(dep);
+  }
+  void setDistance(Dependence dep, unsigned val) { distance[dep] = val; }
+
+  /// The initiation interval (II) is the number of time steps between
+  /// subsequent iterations, i.e. a new iteration is started every II time
+  /// steps. The best possible value is 1, meaning that a corresponding pipeline
+  /// accepts new data every cycle. This property is part of the cyclic
+  /// problem's solution.
+  Optional<unsigned> getInitiationInterval() { return initiationInterval; }
+  void setInitiationInterval(unsigned val) { initiationInterval = val; }
+
+  virtual void clearSolution() override { initiationInterval.reset(); }
+
+protected:
+  virtual LogicalResult verifyDependence(Dependence dep) override;
```

Comment:
Isn't this method already declared on the parent `Problem` class?

Reply:
It is, but I actually want to override it in `CyclicProblem`, as the invariant to be verified is different (albeit very similar):
- *Problem*:
   `startTime(i) + latency(assocOpr(i)) <= startTime(j)`
- *CyclicProblem:*
   `startTime(i) + latency(assocOpr(i)) <= startTime(j) + dist * II`

My original plan was to let the specialised problems selectively override the `checkXYZ()`/`verifyXYZ()` as needed, but I think I need come up with a more flexible solution to reuse this kind of verification code in the future, anyways (things will get interesting once I define the `ModuloProblem` with two base-classes.

Reply:
I don't want to force you to start thinking about a more flexible solution before its needed, I was just asking a dumb question about how overrides work in C++. Is it necessary to declare this method in the header file with `override`? I was thinking you could just override it in the C++ file without having to name it here, but I don't know off the top of my head.